### PR TITLE
[FLINK-6031][yarn] Add config parameter for user-jar inclusion in cla…

### PR DESCRIPTION
**Pull request by zentol**
This PR adds a config parameter to control how user-jars are being handled in regards ot the system class path for per-job yarn clusters.

The parameter allows:
* to disable the inclusion in the system classpath and use the user classpath instead ("DISABLE")
* prepend the user jars to the system class path ("FIRST")
* append the user jars to the system class path ("LAST")
* (default) add the user jars to the system class path based on the lexicographic order ("ORDER")

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 1]
```diff
@@ -200,6 +214,17 @@ public void setTaskManagerMemory(int memoryMb) {
...
+		String configuredUserJarInclusion = flinkConfiguration.getString(YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR);
+		try {
+			userJarInclusion = YarnConfigOptions.UserJarInclusion.valueOf(configuredUserJarInclusion.toUpperCase());
+		} catch (IllegalArgumentException e) {
+			LOG.warn("Configuration parameter {} was configured with an invalid value {}. Falling back to default ({}).",
+				YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.key(),
+				configuredUserJarInclusion,
+				YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.defaultValue());
+			userJarInclusion = YarnConfigOptions.UserJarInclusion.valueOf(YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.defaultValue());
+		}
```
Comment (rmetzger) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: This code and the code above looks like exactly the same.
If this is duplicate code, it should be extracted into one method?

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java [hunk 2]
```diff
@@ -665,58 +685,52 @@ public ApplicationReport startAppMaster(JobGraph jobGraph, YarnClient yarnClient
...
+		String configuredUserJarInclusion = flinkConfiguration.getString(YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR);
+		YarnConfigOptions.UserJarInclusion userJarInclusion;
+		try {
+			userJarInclusion = YarnConfigOptions.UserJarInclusion.valueOf(configuredUserJarInclusion.toUpperCase());
+		} catch (IllegalArgumentException e) {
+			LOG.warn("Configuration parameter {} was configured with an invalid value {}. Falling back to default ({}).",
+				YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.key(),
+				configuredUserJarInclusion,
+				YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.defaultValue());
+			userJarInclusion = YarnConfigOptions.UserJarInclusion.valueOf(YarnConfigOptions.CLASSPATH_INCLUDE_USER_JAR.defaultValue());
```
Comment (rmetzger) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: This also seems very similar to the other duplicates.

File: docs/setup/yarn_setup.md [hunk 3]
```diff
@@ -245,6 +245,18 @@ Note: You can use a different configuration directory per job by setting the env
...
+By default Flink will include the user jars into the system classpath when running a single job. This behavior can be controlled with the `yarn.per-job-cluster.include-job-jar` parameter.
+
+When setting this to `DISABLED` Flink will include the jar in the user classpath instead.
+
+The user-jars position in the class path can be controlled by setting the parameter to one of the following:
+
+- `ORDER`: (default) Adds the jar to the system class path based on the lexicographic order.
+- `FIRST`: Adds the jar to the beginning of the system class path.
+- `LAST`: Adds the jar to the end of the system class path.
+
```
Comment (rmetzger) [inline docs/setup/yarn_setup.md]: Ideally, you put the configuration property also into the  `setup/config.md` page, where all config options are listed.

File: docs/setup/yarn_setup.md [hunk 4]
```diff
@@ -245,6 +245,18 @@ Note: You can use a different configuration directory per job by setting the env
 
 Note: It is possible to combine `-m yarn-cluster` with a detached YARN submission (`-yd`) to "fire and forget" a Flink job to the YARN cluster. In this case, your application will not get any accumulator results or exceptions from the ExecutionEnvironment.execute() call!
 
+### User jars & Classpath
+
+By default Flink will include the user jars into the system classpath when running a single job. This behavior can be controlled with the `yarn.per-job-cluster.include-job-jar` parameter.
```
Comment (rmetzger) [inline docs/setup/yarn_setup.md]: The configuration parameter is actually called `yarn.per-job-cluster.include-user-jar`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java (hunk 2 above)
Comment (zentol) [inline flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java]: this shouldn't be here actually.

**Comment by zentol** [comment]
@rmetzger I've addressed your comments.

**Comment by rmetzger** [comment]
Thank you for addressing my comments so quickly.

The change is good to merge.

**Comment by rmetzger** [comment]
(for both 1.3 and 1.4)

**Comment by rmetzger** [comment]
I'll merge the change now ...

**Pull request closed without merging**